File: merge_sort.rs
fn main() {
    let res = merge_sort(vec![1, 2, 4, 3].as_ref());
    let _res = merge_sort(vec![1, 3, 2, 4, 6, 5, 89, 45].as_ref());
    let _res = merge_sort(vec![1].as_ref());
    let _res = merge_sort(vec![1, 0].as_ref());
    println!("{:?}", res);
}

fn merge_sort(input: &[i32]) -> Vec<i32> {
    let n = input.len();

    if n <= 1 {
        return input.to_vec();
    }

    let (left, right) = input.split_at(n / 2);

    let mut left_i = merge_sort(left).into_iter();
    let mut right_i = merge_sort(right).into_iter();

    let mut res = vec![];
    let mut left = left_i.next();
    let mut right = right_i.next();
    for _ in 0..(left_i.len() + right_i.len() + 2) {
        match (left, right) {
            (Some(l), Some(r)) => {
                if l < r {
                    res.push(l);
                    left = left_i.next();
                } else {
                    res.push(r);
                    right = right_i.next();
                }
            }
            (None, Some(r)) => {
                res.push(r);
                right = right_i.next();
            }
            (Some(l), None) => {
                res.push(l);
                left = left_i.next();
            }

            (None, None) => break,
        }
    }

    res
}

File: merge_sort.go
package main

import "fmt"

func merge(input []int, low int, mid int, high int) {
	// fmt.Println(">>> Left", input[low:mid])
	// fmt.Println(">>> Right", input[mid:high])
	var res = []int{}

	// println(low, mid, high)
	var l, r = low, mid

	for l < mid && r < high {
		if input[l] <= input[r] {
			res = append(res, input[l])
			l = l + 1
		} else {
			res = append(res, input[r])
			r = r + 1
		}
	}

	for l < mid {
		res = append(res, input[l])
		l = l + 1
	}
	for r < high {
		res = append(res, input[r])
		r = r + 1
	}

	var c = 0
	for i := low; i < high; i++ {
		input[i] = res[c]
		c++
	}
	// fmt.Println("::", res)
	// fmt.Println("::", input)
}

func merge_sort(low int, high int, input []int) {
	var mid = (low + high) / 2

	// fmt.Println("For input", input[low:high])
	// fmt.Println("Left", input[low:mid])
	// fmt.Println("Right", input[mid:high])

	if mid == low {
		return
	}

	merge_sort(low, mid, input)
	merge_sort(mid, high, input)

	merge(input, low, mid, high)
}

func main() {
	var input = []int{1, -3, 2, 1, 1, 5, 0, 4, 6, 0, 0, 0}
	merge_sort(0, len(input), input)
	fmt.Println(input)
}
